src/Gui/Document.cpp

```diff
@@ -2531,3 +2531,35 @@ void Document::slotChangePropertyEditor(const App::Document &doc, const App::Pro
         getMainWindow()->setUserSchema(doc.UnitSystem.getValue());
     }
 }
+
+std::vector<App::DocumentObject*> Document::getTreeRootObjects() const

```

Comment:
The implementation works but is very inefficient. In the worst case the runtime will be O(n**2).

Th implementation below is much faster. For testing purposes I used 10,000 elements and it's faster by a factor of 20:
```
    std::vector<App::DocumentObject*> allClaimed;
    std::vector<App::DocumentObject*> docObjects = d->_pcDocument->getObjects();

    for (auto obj : docObjects) {
        ViewProvider* vp = Application::Instance->getViewProvider(obj);
        if (!vp) {
            continue;
        }

        std::vector<App::DocumentObject*> children = vp->claimChildren();
        allClaimed.reserve(allClaimed.size() + children.size());
        for (auto child : children) {
            allClaimed.push_back(child);
        }
    }

    std::sort(allClaimed.begin(), allClaimed.end());
    allClaimed.erase(std::unique(allClaimed.begin(), allClaimed.end()), allClaimed.end());

    std::vector<App::DocumentObject*> rootObjs;
    std::back_insert_iterator< std::vector<App::DocumentObject*> > biit(rootObjs);
    std::set_difference(docObjects.begin(), docObjects.end(),
                        allClaimed.begin(), allClaimed.end(), biit);

    return rootObjs;

```

Reply:
An even faster way could be to use a std::unordered_map of `DocumentObject*, bool ` where by default all boolean values are true. Then for the claimed objects of all view providers the boolean is set to false.

At the end all objects with a boolean of true are top-level objects. Accessing an element of the map is O(1) so that the whole algorithm will be O(n).

Reply:
So it would be this ? 
```
std::vector<App::DocumentObject*> Document::getTreeRootObjects() const {
    std::vector<App::DocumentObject*> docObjects = d->_pcDocument->getObjects();
    std::unordered_map<App::DocumentObject*, bool> objectMap;

    for (auto obj : docObjects) {
        objectMap[obj] = true;
    }

    for (auto obj : docObjects) {
        ViewProvider* vp = Application::Instance->getViewProvider(obj);
        if (!vp) {
            continue;
        }

        std::vector<App::DocumentObject*> children = vp->claimChildren();
        for (auto child : children) {
            objectMap[child] = false;
        }
    }

    std::vector<App::DocumentObject*> rootObjs;
    for (auto& entry : objectMap) {
        if (entry.second) {
            rootObjs.push_back(entry.first);
        }
    }

    return rootObjs;
}
```

Reply:
Did some tests (with 10,000 elements) with the unordered map and this is again faster by a factor of ~30.
```
    std::vector<App::DocumentObject*> docObjects = d->_pcDocument->getObjects();
    std::unordered_map<App::DocumentObject*, bool> rootMap;
    for (auto it : docObjects) {
        rootMap[it] = true;
    }

    for (auto obj : docObjects) {
        ViewProvider* vp = Application::Instance->getViewProvider(obj);
        if (!vp) {
            continue;
        }

        std::vector<App::DocumentObject*> children = vp->claimChildren();
        for (auto child : children) {
            rootMap[child] = false;
        }
    }

    std::vector<App::DocumentObject*> rootObjs;
    for (const auto& it : rootMap) {
        if (it.second) {
            rootObjs.push_back(it.first);
        }
    }
    return rootObjs;
```

Reply:
> So it would be this ?

Yes. It is the final solution now.